General improvements

jgriffiths opened the pull request:
A trial balloon of changes to gauge the reaction. The goal is to eventually move to more idiomatic use of make to express more dependencies and to try to remove the large bits of shell embedded in the makefiles.

The changes to the get-sources and umount targets are examples of this style, where we use make targets to express subtasks instead of fighting make to export variables and hard coding our own iteration in shell.

marmarek commented:
Other than the one comment above it's good, thanks!

jgriffiths commented:
>This would make make chroot-fc23 unmount everything in there. Doesn't looks like a good idea - it's counter intuitive.

It doesn't, because the directory exists so make considers it up to date:
```
[user@personal qubes-builder]$ make chroot-fc23
make: Nothing to be done for 'chroot-fc23'.
```

If you think it is confusing to re-use existing (non-explicit target) directory names as phony action targets as done here, the solution I've seen employed in other places is to name the phony targets after the action, e.g chroot-fc23.umount. In that case `make dirname.umount` would actually unmount it which can be useful for targets like clean.

If you would prefer this approach I'm happy to add it.

marmarek commented:
Yes `make dirname.umount` would make more sense. I think `make dirname` would make sense only when it would actually create/populate that directory (chroot env in this case).

jgriffiths commented:
OK, I've updated and rebased. The impl gets a tiny bit more complicated because it has to generate and strip the pseudo target suffixes. It remains idiomatic.

Following these changes e.g. `make <chrootname>.umount` will umount a given chroot and `make <component>.get-sources'  will call get-sources for <component>. Given that these target names are returned to bash command completion, the latter in particular is much nicer than `COMPONENTS="component" make get-sources` IMO. Once targets like clean/mostlyclean etc are converted this should be a nice and quick way to work with individual components.

jgriffiths commented:
>Actually there is purpose of fetching builder plugins first: to execute its get-sources target before downloading other components. It may for example import some key later used to verify code.
> But it would be good to think of some solution for QubesOS/qubes-issues#1420

There are a couple of separate issues as I see it, revolving around the top level make not having a global view of targets and dependencies. In general its bad form to rely on side effects or ordering unless that dependency is exposed to make. Here we have side effects (key installation and config updates) that can imply new targets and new ordering constraints we cannot know when executing the get-sources target. We also have dependencies hidden from the parent make by virtue of get-sources for components being 'hidden' in recursive make calls. This combination is pretty fatal.

Side effects should generally be fixed by making them explicit targets with correct dependencies. So component_x.get-sources depends on component_y.install_key. These relationships can be dynamically generated by rules easily enough.

If the makefile actions were target driven rather than imperative you would be able to implement get-sources as a pattern rule that produces makefiles. So instead of being phony, make would know that it is fetching/updating a makefile, and would automatically re-read its variables if the makefile was included rather than being used by a recursive make call. That's not really possible with the current impl.

This can be fixed by refactoring recursion into inclusion for some targets like source fetching, cleaning etc. But to do that requires coordinated updates to all of the qubes-src/ components including the ones maintained by others. That will take quite some time no doubt.

In the meantime I think you have to accept that a single call from the parent make will not work. You can possibly force make to call itself recursively and quit after installing each builder, but that is a pretty grim 'fix' and usually flaky. Or you can have make quit with a message to re-run when we detect that something has changed that requires re-reading.

We can also separate the fetching of sources and their validation. This should be done anyway in my view since the dependencies here are all screwed up e.g. downloaded the sources are re-validated needlessly when the files haven't changed, for example.

I will experiment with making the non-builder source fetching depend on the builders source fetching, which would cause them to be download first. I don't think that solves anything (i.e. I don't think the current code solves anything) on its own though.

marmarek commented:
> In the meantime I think you have to accept that a single call from the parent make will not work.

Fair enough.

jgriffiths commented:
> BTW Script to build template in DispVM explicitly calls make get-sources

Hmm, I think this script has bit-rotted somewhat. at least parseConfigLocation() is broken, its calling ./get-sources.sh which no longer exists.

I could test and fix this and other changes much faster if I could cache the dnf fetches when building the chroot. Is there any way to do that or otherwise proxy the dnf calls to avoid re-downloading the rpms after every clean build?

marmarek commented:
On Wed, Mar 09, 2016 at 05:07:13PM -0800, jgriffiths wrote:
> > BTW Script to build template in DispVM explicitly calls make get-sources
> 
> Hmm, I think this script has bit-rotted somewhat. at least parseConfigLocation() is broken, its calling ./get-sources.sh which no longer exists.

Indeed, the online case probably is broken now. But the local case works
just fine :)

> I could test and fix this and other changes much faster if I could cache the dnf fetches when building the chroot. Is there any way to do that or otherwise proxy the dnf calls to avoid re-downloading the rpms after every clean build?

Don't know... It was possible with yum with `--setopt=cachedir=...
--setopt=keepcache=1`, but I don't see cachedir option in dnf.conf
documentation anymore. Probably one of many features dropped in dnf...

But instead of caching by dnf, you can setup caching proxy and use
REPO_PROXY builder.conf option.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?
